components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+        ui.status(Status::Verifying, artifact.ident()?)?;
+        self.verify_artifact(ui, package, &mut artifact)?;
+        Ok(artifact)
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Retrieve the identified package from the depot, ensuring that
+    /// the artifact is cached locally.
+    fn fetch_artifact<T>(&self, ui: &mut T, package: &PackageIdentTarget) -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading, package)?;
+        match self.api_client
+                  .fetch_package((&package.ident, package.target),
+                                 self.token,
+                                 self.artifact_cache_path,
+                                 ui.progress())
+        {
+            Ok(_) => Ok(()),
+            Err(api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                println!("Host platform or architecture not supported by the targeted depot; \
+                          skipping.");
+                Ok(())
+            }
+            Err(e) => Err(Error::from(e)),
+        }
+    }
+
+    fn fetch_origin_key<T>(&self,
+                           ui: &mut T,
+                           name_with_rev: &str,
+                           token: Option<&str>)
+                           -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading,
+                  format!("{} public origin key", &name_with_rev))?;
+        let (name, rev) = parse_name_with_rev(&name_with_rev)?;
+        self.api_client
+            .fetch_origin_key(&name, &rev, token, self.key_cache_path, ui.progress())?;
+        ui.status(Status::Cached,
+                  format!("{} public origin key", &name_with_rev))?;
+        Ok(())
+    }
+
+    fn verify_artifact<T>(&self,
+                          ui: &mut T,
+                          package: &PackageIdentTarget,
+                          artifact: &mut PackageArchive)
+                          -> Result<()>
+        where T: UIWriter
+    {
+        let artifact_ident = artifact.ident()?;
+        if package.ident.as_ref() != &artifact_ident {
+            return Err(Error::ArtifactIdentMismatch((artifact.file_name(),
+                                                     artifact_ident.to_string(),
+                                                     package.to_string())));
+        }
+
+        let artifact_target = artifact.target()?;
+        if package.target != artifact_target {
+            return Err(Error::HabitatCore(hcore::Error::WrongActivePackageTarget(
+                package.target,
+                artifact_target,
+            )));
+        }
+
+        let nwr = artifact::artifact_signer(&artifact.path)?;
+        if SigKeyPair::get_public_key_path(&nwr, self.key_cache_path).is_err() {
+            self.fetch_origin_key(ui, &nwr, self.token)?;
```

Comment:
We don't want to be pulling keys by cracking open each archive - that is prohibitively expensive.
Instead, we should just download all the public keys for the origin(s) into the cache (check and download if they are not already in the cache).